test-data/unit/check-classes.test

```diff
@@ -894,6 +892,281 @@ a.f = a.f # E: Property "f" defined in "A" is read-only
 a.f.x # E: "int" has no attribute "x"
 [builtins fixtures/property.pyi]
 
+-- Descriptors
+-- -----------
+
+
+[case testAccessingNonDataDescriptor]
+class D():
+    def __get__(self, instance: Any, owner: Any) -> str:
+        return 'some string'
+
+class A:
+    f = D()
+
+a = A() # type: A
+reveal_type(a.f)  # E: Revealed type is 'builtins.str'
+
+[case testSettingDataDescriptor]
+class D():
+    def __get__(self, instance: Any, owner: Any) -> str:
+        return 'some string'
+
+    def __set__(self, instance: Any, value: str) -> None:
+        pass
+
+class A:
+    f = D()
+
+a = A() # type: A
+a.f = ''
+a.f = 1 # E: Incompatible types in assignment (expression has type "int", variable has type "str")
+
+[case testReadingDescriptorWithoutDunderGet]
+from typing import Union
+
+class D():
+    def __set__(self, instance: Any, value: str) -> None:
+        pass
+
+class A:
+    f = D()  # type: Union[D, str]
+
+    def __init__(self):
+        self.f = 'some string'  # type: str
+
+a = A() # type: A
+reveal_type(a.f)  # E: Revealed type is 'Union[__main__.D, builtins.str]'
+
+[case testAccessingGenericNonDataDescriptor]
+from typing import TypeVar, Type, Generic
+
+T = TypeVar('T')
+V = TypeVar('V')
+class D(Generic[T, V]):
+    def __init__(self, value: V) -> None:
+        self.value = value
+    def __get__(self, instance: T, owner: Type[T]) -> V:
```

Comment:
Hm. I'd like to see a test where D is generic in V but not in T first. And I'd somehow hoped that the type declarations inside class A below would not be needed.